### PR TITLE
Add bitset optimization described in #664

### PR DESCRIPTION
**Pull request by tmfink**
Implements optimization described in #664.

### CODE REVIEW

**Comment by fitzgen** [comment]
Thanks @tmfink -- taking a look now

File: src/ir/named.rs [hunk 1]
```diff
@@ -152,7 +153,7 @@ pub trait MonotoneFramework: Sized + fmt::Debug {
     ///
     /// This is just generic (and not `ItemId`) so that we can easily unit test
     /// without constructing real `Item`s and their `ItemId`s.
-    type Node: Copy;
+    type Node: Copy + Into<usize>;
```
Comment (fitzgen) [inline src/ir/named.rs]: Nice.

File: src/ir/named.rs [hunk 2]
```diff
@@ -198,10 +199,24 @@ pub fn analyze<Analysis>(extra: Analysis::Extra) -> Analysis::Output
     let mut analysis = Analysis::new(extra);
     let mut worklist = analysis.initial_worklist();
 
+    // INVARIANT: node_ids_in_worklist.contains(node.id) if and only if
+    // worklist.contains(node)
+    let mut node_ids_in_worklist = BitSet::new();
```
Comment (fitzgen) [inline src/ir/named.rs]: Nice comment :)

File: src/ir/named.rs [hunk 3]
```diff
@@ -198,10 +199,24 @@ pub fn analyze<Analysis>(extra: Analysis::Extra) -> Analysis::Output
...
+
     while let Some(node) = worklist.pop() {
+        // We just popped `node` from `worklist`, so remove `node` from
+        // `node_ids_in_worklist`
+        node_ids_in_worklist.remove(node.into());
```
Comment (fitzgen) [inline src/ir/named.rs]: Can we `debug_assert!` that the node was in the worklist? It's good to do our best to make sure our invariants are upheld in the face of future changes.

**Comment by fitzgen** [comment]
I measured the time spent in the `analyze` function with and without the optimization here, 50 times each, and crunched the results using the p awesome `histogram.py` from https://github.com/bitly/data_hacks:

```
$ histogram.py < with-bit-set.csv
# NumSamples = 50; Min = 24849028.00; Max = 256668384.00
# Mean = 111103071.160000; Variance = 3115017899422295.000000; SD = 55812345.403345; Median 107741588.000000
# each ∎ represents a count of 1
  24849028.0000 - 48030963.6000 [     5]: ∎∎∎∎∎
  48030963.6000 - 71212899.2000 [     9]: ∎∎∎∎∎∎∎∎∎
  71212899.2000 - 94394834.8000 [     5]: ∎∎∎∎∎
 94394834.8000 - 117576770.4000 [    11]: ∎∎∎∎∎∎∎∎∎∎∎
117576770.4000 - 140758706.0000 [     8]: ∎∎∎∎∎∎∎∎
140758706.0000 - 163940641.6000 [     4]: ∎∎∎∎
163940641.6000 - 187122577.2000 [     4]: ∎∎∎∎
187122577.2000 - 210304512.8000 [     0]: 
210304512.8000 - 233486448.4000 [     2]: ∎∎
233486448.4000 - 256668384.0000 [     2]: ∎∎

$ histogram.py < without-bit-set.csv
# NumSamples = 50; Min = 24666320.00; Max = 212343031.00
# Mean = 90409389.380000; Variance = 2379900611247277.500000; SD = 48784225.024564; Median 78896049.000000
# each ∎ represents a count of 1
  24666320.0000 - 43433991.1000 [     7]: ∎∎∎∎∎∎∎
  43433991.1000 - 62201662.2000 [    10]: ∎∎∎∎∎∎∎∎∎∎
  62201662.2000 - 80969333.3000 [    10]: ∎∎∎∎∎∎∎∎∎∎
  80969333.3000 - 99737004.4000 [     6]: ∎∎∎∎∎∎
 99737004.4000 - 118504675.5000 [     6]: ∎∎∎∎∎∎
118504675.5000 - 137272346.6000 [     3]: ∎∎∎
137272346.6000 - 156040017.7000 [     2]: ∎∎
156040017.7000 - 174807688.8000 [     1]: ∎
174807688.8000 - 193575359.9000 [     2]: ∎∎
193575359.9000 - 212343031.0000 [     3]: ∎∎∎
```

Consider me very surprised: it looks like the bit set "optimization" doesn't pay for itself.

I guess we can close this PR and the related issue.

Thanks for investigating this @tmfink! Too bad it didn't end up paying off this time. Are you interested in hacking on another bindgen bug? If so, I have some potential issues in mind. Let me know!

**Comment by tmfink** [comment]
@fitzgen I'm up for tackling another bug; I'm particularly interested #699 since it would help me with https://github.com/tmfink/capstone-sys/tree/bindgen-wip

**Comment by fitzgen** [comment]
@tmfink yeah, that sounds great! I think @emilio might be a slightly better mentor for that issue than me, since he is generally more familiar with the codegen phase. But I can try and answer any questions that you might come up with.

https://github.com/servo/rust-bindgen/blob/cabfac71fc454196656ff63e5cc761294896831d/src/codegen/mod.rs#L2109 is where we detect constified enums during codegen, and if you look at how that variable is used after, that should lead you to all the relevant bits of code.

I'll go ahead and assign you to that issue. Good luck!

**Pull request closed without merging**